opni-insights-service/app/main.py

```diff
@@ -449,6 +450,113 @@ async def get_anomalies_breakdown(start_ts, end_ts):
         return anomaly_breakdown_dict
     return anomaly_breakdown_dict
 
+async def get_areas_of_interest(start_ts: int, end_ts: int):
+    '''
+    This function will fetch all areas of interest between the starting and ending timestamp. It will return a list
+    of dictionary objects with a start_ts and end_ts of each of the overarching areas of interest and a list consisting
+    of the smaller areas of interest that are present within this expanded time interval. The idea behind this
+    is to reduce the immediate number of area of interest time intervals that are returned.
+    '''
+    logging.info(
+        f"Received request to obtain all areas of interest between {start_ts} and {end_ts}"
+    )
+
+    query = {
+        "query": {
+            "bool": {
+                "must": [{"match": {"anomaly_level.keyword": "Anomaly"}}],
+                "filter": [{"range": {"timestamp": {"gte": start_ts, "lte": end_ts}}}],
+            }
+        },
+        "aggs": {
+            "logs_over_time": {
+                "date_histogram": {"field": "timestamp", "calendar_interval": "minute"},
+                "aggs": {"total_normal": {"sum": {"field": "anomaly_predicted_count"}}},
+            }
+        },
+    }
+    try:
+        res = await es_instance.search(index="logs", body=query, size=0)
+        df = pd.DataFrame(res["aggregations"]["logs_over_time"]["buckets"])
+        areas_of_interest = []
+        if len(df) > 0:
+            # anomaly_predicted_count == 2 for 'Anomaly' labeled log messages
+            df["doc_count"] = df["doc_count"] / 2
+            df["moving_avg"] = df["doc_count"].rolling(60).mean() * 1.3
+            df["AOI"] = df.doc_count.gt(df.moving_avg.shift())
+            areas_of_interest_df = df[df.AOI == True]
+            areas_of_interest_df["time_diff"] = areas_of_interest_df["key"].diff()
+            current_aoi_start = -1
+            prev_segment_start = -1
+            aoi_start_ts = -1
+            current_interval_areas = []
+            '''
+            Iterate over the areas of interest found and create a larger area of interest time interval which will contain
+            time intervals which are within 10 minutes of the previous area of interest time interval. This reduces
+            the outward number of areas of interest time intervals returned.
+            '''
+            for index, row in areas_of_interest_df.iterrows():
+                if current_aoi_start == -1:
+                    current_aoi_start = row["key"]
+                    aoi_start_ts = row["key"]
+                if not pd.isna(row['time_diff']) and row["time_diff"] != MILLISECONDS_MINUTE:
+                    if row["key"] - prev_segment_start > (11 * MILLISECONDS_MINUTE):
```

Comment:
can we add docstring for this magic number `11`?

Reply:
added documentation and made some changes that should make this clearer.